homeassistant/components/envisalink.py

```diff
@@ -0,0 +1,198 @@
+"""
+Support for Envisalink devices.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/envisalink/
+"""
+import logging
+import time
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.discovery import load_platform
+from homeassistant.util import convert
+
+REQUIREMENTS = ['pyenvisalink==0.9', 'pydispatcher==2.0.5']
+
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = 'envisalink'
+
+EVL_CONTROLLER = None
+
+CONF_EVL_HOST = 'host'
+CONF_EVL_PORT = 'port'
+CONF_PANEL_TYPE = 'panel_type'
+CONF_EVL_VERSION = 'evl_version'
+CONF_CODE = 'code'
+CONF_USERNAME = 'user_name'
+CONF_PASS = 'password'
+CONF_EVL_KEEPALIVE = 'keepalive_interval'
+CONF_ZONEDUMP_INTERVAL = 'zonedump_interval'
+CONF_ZONES = 'zones'
+CONF_PARTITIONS = 'partitions'
+
+DEFAULT_PORT = 4025
+DEFAULT_EVL_VERSION = 3
+DEFAULT_KEEPALIVE = 60
+DEFAULT_ZONEDUMP_INTERVAL = 30
+
+SIGNAL_ZONE_UPDATE = 'zones_updated'
+SIGNAL_PARTITION_UPDATE = 'partition_updated'
+SIGNAL_KEYPAD_UPDATE = 'keypad_updated'
+
+REQ_VAR_ERR = "Missing required variable: {0}"
+INVALID_PANEL_ERR = REQ_VAR_ERR + " Valid values are HONEYWELL/DSC."
+
+
+# pylint: disable=unused-argument, too-many-function-args, too-many-locals
+# pylint: disable=too-many-return-statements
+def setup(hass, base_config):
+    """Common setup for Envisalink devices."""
+    from pyenvisalink import EnvisalinkAlarmPanel
+    from pydispatch import dispatcher
+
+    global EVL_CONTROLLER
+
+    config = base_config.get(DOMAIN)
+
+    if config.get(CONF_EVL_HOST) is None:
```

Comment:
Instead of doing all these error checks, you could consider using a voluptuous schema like for example MQTT [uses](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/__init__.py#L87-L107). Naming it `CONFIG_SCHEMA` will trigger Home Assistant to validate your config before it arrives in `setup(hass, config)`